examples/amm/src/contract.rs

```diff
@@ -329,25 +363,132 @@ impl Amm {
+                .expect("Couldn't calculate_output_amount");
+
+            if output_amount >= swap_bid.minimum_output_token_amount {
+                // Execute swap
+                self.receive_from_account(
+                    swap_bid.owner,
+                    swap_bid.input_token_idx,
+                    swap_bid.input_amount,
+                )
+                .await
+                .expect("Couldn't receive_from_account");
+                self.send_to(swap_bid.owner, swap_bid.output_token_idx, output_amount)
+                    .await
+                    .expect("Couldn't send_to");
+
+                swap_bid_ids_to_remove.push(swap_bid_id);
+            }
+        }
+
+        for swap_bid_id in swap_bid_ids_to_remove {
+            self.swap_bids.remove(&swap_bid_id)?;
+        }
+
+        Ok(())
+    }
+
+    fn calculate_output_amount(
+        &mut self,
+        input_amount: Amount,
+        input_pool_balance: Amount,
+        output_pool_balance: Amount,
+    ) -> Result<Amount, AmmError> {
+        if input_pool_balance == Amount::ZERO || output_pool_balance == Amount::ZERO {
+            return Err(AmmError::InvalidPoolBalanceError);
+        }
+
+        let input_amount_bigint = BigUint::from_u128(u128::from(input_amount))
+            .expect("Couldn't generate input_amount in bigint");
+        let output_pool_balance_bigint = BigUint::from_u128(u128::from(output_pool_balance))
+            .expect("Couldn't generate output_pool_balance in bigint");
+        let input_pool_balance_bigint = BigUint::from_u128(u128::from(input_pool_balance))
+            .expect("Couldn't generate input_pool_balance in bigint");
+
+        // Logic for this is the following:
+        // This is a Constant Product Automated Market Maker, or CPAMM, so we want
+        // the product to remain constant.
+        // That means that this is the equation we need to solve to find output_amount:
+        //      (input_pool_balance + input_amount) * (output_pool_balance - output_amount) = input_pool_balance * output_pool_balance
+        //      output_pool_balance - output_amount = (input_pool_balance * output_pool_balance) / (input_pool_balance + input_amount)
+        //      output_amount = output_pool_balance - (input_pool_balance * output_pool_balance) / (input_pool_balance + input_amount)
+        //      output_amount = (output_pool_balance * (input_pool_balance + input_amount) - (input_pool_balance * output_pool_balance)) / (input_pool_balance + input_amount)
+        //      output_amount = (input_pool_balance * output_pool_balance + input_amount * output_pool_balance - input_pool_balance * output_pool_balance) / (input_pool_balance + input_amount)
+        //      output_amount = (input_amount * output_pool_balance) / (input_pool_balance + input_amount)
+
+        // Numerator will be a number with 36 decimal points here
+        let numerator_bigint = input_amount_bigint.clone() * output_pool_balance_bigint;
```

Comment:
That clone should be unnecessary. (Maybe you need to make the `&output_pool_balance_bigint` a reference for that.)